packages/server/lib/xhr_ws_server.ts

```diff
@@ -0,0 +1,70 @@
+import _ from 'lodash'
+import Bluebird from 'bluebird'
+import debugModule from 'debug'
+
+const debug = debugModule('cypress:server:xhr_ws_server')
+
+function trunc (str) {
+  return _.truncate(str, {
+    length: 100,
+    omission: '... [truncated to 100 chars]',
+  })
+}
+
+export function create () {
+  let incomingXhrs: any = {}
+
+  function onIncomingXhr (id, data) {
+    debug('onIncomingXhr %o', { id, res: trunc(data) })
+    const deferred = incomingXhrs[id]
+
+    if (deferred) {
+      return deferred.resolve({
+        data,
+      })
+    }
+
+    incomingXhrs[id] = data
+  }
+
+  function getDeferredResponse (id) {
+    debug('getDeferredResponse %o', { id })
+    // if we already have it, send it
+    const res = incomingXhrs[id]
+
+    if (res) {
+      if (res.then) {
+        debug('returning existing deferred promise for %o', { id })
+      }
+
+      debug('already have deferred response %o', { id, res: trunc(res) })
+      delete incomingXhrs[id]
+
+      return res
+    }
+
+    return new Bluebird((resolve, reject) => {
+      debug('do not have response, waiting %o', { id })
+      incomingXhrs[id] = { resolve, reject }
+    })
+    .tap((res) => {
+      debug('deferred response found %o', { id, res: trunc(res) })
+    })
+  }
+
+  function reset () {
+    debug('resetting incomingXhrs %o', { length: incomingXhrs.length })
+
+    _.forEach(incomingXhrs, ({ reject }) => {
+      reject(new Error('This stubbed XHR was pending on a stub response object from the driver, but the test ended before that happened.'))
```

Comment:
this does not feel correct - tests can end while the XHR is in flight without that being an actual error

Reply:
This is really just to ensure that the HTTP request is closed before creating new tests; otherwise, the connections would pile up until they timed out.

It could also be fulfilled with a successful response, or left to time out.